docs/device/config.md
# Device Configuration

Device configuration can be broken down into required and optional configuration:

## Required Configuration
- `identity.orgId` Your organization ID.
- `identity.typeId` The type of the device. Think of the device type is analagous to a model number.
- `identity.deviceId` A unique ID to identify a device. Think of the device id as analagous to a serial number.
- `auth.token` An authentication token to securely connect your device to Watson IoT Platform.

## Optional Configuration
- `options.domain` A boolean value indicating which Watson IoT Platform domain to connect to (e.g. if you have a dedicated platform instance). Defaults to `internetofthings.ibmcloud.com`
- `options.logLevel` Controls the level of logging in the client, can be set to `error`, `warning`, `info`, or `debug`.  Defaults to `info`.
- `options.mqtt.port` A integer value defining the MQTT port.  Defaults to auto-negotiation.
- `options.mqtt.transport` The transport to use for MQTT connectivity - `tcp` or `websockets`.
- `options.mqtt.cleanStart` A boolean value indicating whether to discard any previous state when reconnecting to the service.  Defaults to `False`.
- `options.mqtt.sessionExpiry` When cleanStart is disabled, defines the maximum age of the previous session (in seconds).  Defaults to `False`.
- `options.mqtt.keepAlive` Control the frequency of MQTT keep alive packets (in seconds).  Details to `60`.
- `options.mqtt.caFile` A String value indicating the path to a CA file (in pem format) to use in verifying the server certificate.  Defaults to `messaging.pem` inside this module. Use the special string `"_os_"` to use default python/OS truststore.


The config parameter when constructing an instance of `wiotp.sdk.device.DeviceClient` expects to be passed a dictionary containing this configuration:

```python
myConfig = { 
    "identity": {
        "orgId": "org1id",
        "typeId": "raspberry-pi-3"
        "deviceId": "00ef08ac05"
    }.
    "auth" {
        "token": "Ab$76s)asj8_s5"
    },
    "options": {
        "domain": "internetofthings.ibmcloud.com",
        "logLevel": "error|warning|info|debug",
        "mqtt": {
            "port": 8883,
            "transport": "tcp|websockets",
            "cleanStart": True|False,
            "sessionExpiry": 3600,
            "keepAlive": 60,
            "caFile": "/path/to/certificateAuthorityFile.pem"
        }
    }
}
client = wiotp.sdk.device.DeviceClient(config=myConfig, logHandlers=None)
```

In most cases you will not manually build the `config` dictionary.  Two helper methods are provided to make configuration simple:


## YAML File Support

`wiotp.sdk.device.parseConfigFile()` allows one to easily pass in device configuration from environment variables.

```python
import wiotp.sdk.device

myConfig = wiotp.sdk.device.parseConfigFile("device.yaml")
client = wiotp.sdk.device.DeviceClient(config=myConfig, logHandlers=None)
```

### Minimal Required Configuration File

```yaml
identity:
    orgId: org1id
    typeId: raspberry-pi
    deviceId: 00ef08ac05
auth:
    token: Ab$76s)asj8_s5
```

### Complete Configuration File

This file defines all optional configuration parameters.

```yaml
identity:
    orgId: org1id
    typeId: raspberry-pi
    deviceId: 00ef08ac05
auth:
    token: Ab$76s)asj8_s5
options:
    domain: internetofthings.ibmcloud.com
    logLevel: debug
    mqtt:
        port: 8883
        transport: tcp
        cleanStart: true
        sessionExpiry: 7200
        keepAlive: 120
        caFile: /path/to/certificateAuthorityFile.pem
```


## Environment Variable Support

`wiotp.sdk.device.parseEnvVars()` allows one to easily pass in device configuration from environment variables.

```python
import wiotp.sdk.device

myConfig = wiotp.sdk.device.parseEnvVars()
client = wiotp.sdk.device.DeviceClient(config=myConfig, logHandlers=None)
```

### Minimal Required Environment Variables
- `WIOTP_IDENTITY_ORGID`
- `WIOTP_IDENTITY_TYPEID`
- `WIOTP_IDENTITY_DEVICEID`
- `WIOTP_AUTH_TOKEN`

### Optional Additional Environment Variables
- `WIOTP_OPTIONS_DOMAIN`
- `WIOTP_OPTIONS_LOGLEVEL`
- `WIOTP_OPTIONS_MQTT_PORT`
- `WIOTP_OPTIONS_MQTT_TRANSPORT`
- `WIOTP_OPTIONS_MQTT_CAFILE`
- `WIOTP_OPTIONS_MQTT_CLEANSTART`
- `WIOTP_OPTIONS_MQTT_SESSIONEXPIRY`
- `WIOTP_OPTIONS_MQTT_KEEPALIVE`



docs/gateway/config.md
# Gateway Configuration

Gateway configuration can be broken down into required and optional configuration:

## Required Configuration
- `identity.orgId` Your organization ID.
- `identity.typeId` The type of the device. Think of the device type is analagous to a model number.
- `identity.deviceId` A unique ID to identify a device. Think of the device id as analagous to a serial number.
- `auth.token` An authentication token to securely connect your device to Watson IoT Platform.

## Optional Configuration
- `options.domain` A boolean value indicating which Watson IoT Platform domain to connect to (e.g. if you have a dedicated platform instance). Defaults to `internetofthings.ibmcloud.com`
- `options.logLevel` Controls the level of logging in the client, can be set to `error`, `warning`, `info`, or `debug`.  Defaults to `info`.
- `options.mqtt.port` A integer value defining the MQTT port.  Defaults to auto-negotiation.
- `options.mqtt.transport` The transport to use for MQTT connectivity - `tcp` or `websockets`.
- `options.mqtt.cleanStart` A boolean value indicating whether to discard any previous state when reconnecting to the service.  Defaults to `False`.
- `options.mqtt.sessionExpiry` When cleanStart is disabled, defines the maximum age of the previous session (in seconds).  Defaults to `False`.
- `options.mqtt.keepAlive` Control the frequency of MQTT keep alive packets (in seconds).  Details to `60`.
- `options.mqtt.caFile` A String value indicating the path to a CA file (in pem format) to use in verifying the server certificate.  Defaults to `messaging.pem` inside this module. Use the special string `"_os_"` to use default python/OS truststore.


The config parameter when constructing an instance of `wiotp.sdk.gateway.GatewayClient` expects to be passed a dictionary containing this configuration:

```python
myConfig = { 
    "identity": {
        "orgId": "org1id",
        "typeId": "raspberry-pi-3"
        "deviceId": "00ef08ac05"
    }.
    "auth" {
        "token": "Ab$76s)asj8_s5"
    },
    "options": {
        "domain": "internetofthings.ibmcloud.com",
        "logLevel": "error|warning|info|debug",
        "mqtt": {
            "port": 8883,
            "transport": "tcp|websockets",
            "cleanStart": True|False,
            "sessionExpiry": 3600,
            "keepAlive": 60,
            "caFile": "/path/to/certificateAuthorityFile.pem"
        }
    }
}
client = wiotp.sdk.gateway.GatewayClient(config=myConfig, logHandlers=None)
```

In most cases you will not manually build the `config` dictionary.  Two helper methods are provided to make configuration simple:


## YAML File Support

`wiotp.sdk.gateway.parseConfigFile()` allows one to easily pass in gateway configuration from environment variables.

```python
import wiotp.gateway.sdk

myConfig = wiotp.sdk.device.parseConfigFile("gateway.yaml")
client = ibmiotf.gateway.Client(config=myConfig, logHandlers=None)
```

### Minimal Required Configuration File

```yaml
identity:
    orgId: org1id
    typeId: raspberry-pi
    deviceId: 00ef08ac05
auth:
    token: Ab$76s)asj8_s5
```

### Complete Configuration File

This file defines all optional configuration parameters.

```yaml
identity:
    orgId: org1id
    typeId: raspberry-pi
    deviceId: 00ef08ac05
auth:
    token: Ab$76s)asj8_s5
options:
    domain: internetofthings.ibmcloud.com
    logLevel: debug
    mqtt:
        port: 8883
        transport: tcp
        cleanStart: true
        sessionExpiry: 7200
        keepAlive: 120
        caFile: /path/to/certificateAuthorityFile.pem
```


## Environment Variable Support

`wiotp.sdk.gateway.parseEnvVars()` allows one to easily pass in gateway configuration from environment variables.

```python
import wiotp.sdk.gateway

myConfig = wiotp.sdk.gateway.parseEnvVars()
client = wiopt.sdk.gateway.Client(config=myConfig, logHandlers=None)
```

### Minimal Required Environment Variables
- `WIOTP_IDENTITY_ORGID`
- `WIOTP_IDENTITY_TYPEID`
- `WIOTP_IDENTITY_DEVICEID`
- `WIOTP_AUTH_TOKEN`

### Optional Additional Environment Variables
- `WIOTP_OPTIONS_DOMAIN`
- `WIOTP_OPTIONS_LOGLEVEL`
- `WIOTP_OPTIONS_MQTT_PORT`
- `WIOTP_OPTIONS_MQTT_TRANSPORT`
- `WIOTP_OPTIONS_MQTT_CAFILE`
- `WIOTP_OPTIONS_MQTT_CLEANSTART`
- `WIOTP_OPTIONS_MQTT_SESSIONEXPIRY`
- `WIOTP_OPTIONS_MQTT_KEEPALIVE`



CHANGES.txt
v0.3.0
======
- Code Consistency:
  - Renamed ibmiotf.gateway.ManagedGateway to ibmiotf.gateway.ManagedClient
  - Renamed ibmiotf.gateway.Client.apiClient to ibmiotf.gateway.Client.api
- Removed unnecessary api.addGatewayDeviceType() method.  Use addDeviceType(typeId, classId="Gateway") which this was a duplicate of
- Removed unnecessary api.registerDeviceUnderGateway() method.  Use addDevice(typeId, deviceId) where typeId is the ID of a Gateway type
- Removed api.getDevicesConnectedThroughGateway() method.  It returns a list of gateways of the specified type.
  - Use api.getDevicesForType() or api.getDevice(), which this method was a duplicate of depending on the options provided to it.

v0.2.7
======
 - Add device schema and state APIs

v0.2.6
======
 - Support for "play" organization
 - Independent HttpClient for application and device
 - Switch to new URL for HTTP messaging API
 - Clean up lingering references to sunset historian service in samples

v0.2.5
======
 - TDB

v0.2.4
======
- Introduced support for dedicated and local offerings (via 'domain' configuration option)

v0.2.3
======
- Fixed bug in application re-subscribe logic on connect

v0.2.2
======
- Application client remembers & automatically restores subscriptions on disconnect/reconnect

v0.2.1
======
- Application client only requires api key and token
- Update to last event cache support 

v0.2.0
======
- The library now uses typeId, deviceId, and eventId consistently
- Initial version of last event cache support 
- Changes to API support:
  - Mixed use of queryParameters & parameters consolidated to always use parameters retrieveDevices() & getAllDevices() removed. Single getDevices() method remains
  - addMultipleDevices renamed to registerDevices() for consistency with registerDevice() method
  - getAllDeviceTypes() renamed to getDeviceTypes() for consistency with other getResourceTypePlural methods 
  - IoTFCReSTException now APIException
  - getDeviceConnectionLogs() renamed getConnectionLogs & restructured to support parameters object instead of deviceTypeId and deviceId

v0.1.8
======

- Bug fix for application client command publish
- Add raw payload property to Command object passed to application callback handlers

v0.1.7
======

- Bug fix for application client event publish

v0.1.6
======

- Add override support for verification of API calls

v0.1.5
======

- 

v0.1.4
======

- Resolves issue with __subscribeToCommands() in device.Client introduced in 0.1.3

v0.1.3
======

- 

v0.1.2
======

- Bug fixes for messageEncoderModule management & logAndRaiseException mangling


v0.1.1
======

- Fix packaging errors in 0.1.0


v0.1
======

- Support for custom message formats & parsers ("json" and "json-iotf" supported out-of-the-box)
- Switch to RotatingFileHandler for logging
- Tune down default log level to INFO


v0.0.9
======

- Support for MQTT over SSL (requires TLS 1.2 support in Python 3)
- Support for HTTP API calls from application client


v0.0.7
======

- Removed auto-connect in client constructor


v0.0.7
======

- Update to support paho-mqtt 1.0 release


v0.0.5
======

- First public (pre-)release


src/wiotp/sdk/application/config.py
# *****************************************************************************
# Copyright (c) 2014, 2024 IBM Corporation and other Contributors.
#
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v10.html
# *****************************************************************************

from collections import defaultdict
import os
import yaml
import logging
import uuid

from wiotp.sdk import ConfigurationException


class ApplicationClientConfig(defaultdict):
    def __init__(self, **kwargs):
        # previously quickstart supported unauthenticated connections but has been removed
        if "auth" not in kwargs:
            raise ConfigurationException("Missing auth from configuration")
        if "key" not in kwargs["auth"] or kwargs["auth"]["key"] is None:
            raise ConfigurationException("Missing auth.key from configuration")
        if "token" not in kwargs["auth"] or kwargs["auth"]["token"] is None:
            raise ConfigurationException("Missing auth.token from configuration")

        if "options" in kwargs and "mqtt" in kwargs["options"]:
            if "port" in kwargs["options"]["mqtt"] and kwargs["options"]["mqtt"]["port"] is not None:
                if not isinstance(kwargs["options"]["mqtt"]["port"], int):
                    raise ConfigurationException("Optional setting options.port must be a number if provided")
            if "cleanSession" in kwargs["options"]["mqtt"] and not isinstance(
                kwargs["options"]["mqtt"]["cleanSession"], bool
            ):
                raise ConfigurationException("Optional setting options.cleanSession must be a boolean if provided")

        # Set defaults for optional configuration
        if "identity" not in kwargs:
            kwargs["identity"] = {}

        if "appId" not in kwargs["identity"]:
            kwargs["identity"]["appId"] = str(uuid.uuid4())

        if "options" not in kwargs:
            kwargs["options"] = {}

        if "domain" not in kwargs["options"] or kwargs["options"]["domain"] is None:
            kwargs["options"]["domain"] = "internetofthings.ibmcloud.com"

        if "logLevel" not in kwargs["options"] or kwargs["options"]["logLevel"] is None:
            kwargs["options"]["logLevel"] = logging.INFO

        if "mqtt" not in kwargs["options"]:
            kwargs["options"]["mqtt"] = {}

        if "instanceId" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["instanceId"] = None

        if "port" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["port"] = None

        if "transport" not in kwargs["options"]["mqtt"] or kwargs["options"]["mqtt"]["transport"] is None:
            kwargs["options"]["mqtt"]["transport"] = "tcp"

        if "cleanStart" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["cleanStart"] = False

        if "sessionExpiry" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["sessionExpiry"] = 3600

        if "keepAlive" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["keepAlive"] = 60

        if "caFile" not in kwargs["options"]["mqtt"]:
            kwargs["options"]["mqtt"]["caFile"] = None

        if "http" not in kwargs["options"]:
            kwargs["options"]["http"] = {}

        if "verify" not in kwargs["options"]["http"] or kwargs["options"]["http"]["verify"] is None:
            kwargs["options"]["http"]["verify"] = True

        dict.__init__(self, **kwargs)

    @property
    def orgId(self):
        # Get the orgId from the apikey (format: a-orgid-randomness)
        return self.apiKey.split("-")[1]

    @property
    def appId(self):
        return self["identity"]["appId"]

    @property
    def instanceId(self):
        return self["options"]["mqtt"]["instanceId"]

    @property
    def clientId(self):
        if self.instanceId is None:
            return "a:%s:%s" % (self.orgId, self.appId)
        else:
            return "A:%s:%s:%s" % (self.orgId, self.appId, self.instanceId)

    @property
    def apiKey(self):
        return self["auth"]["key"] if ("auth" in self) else None

    @property
    def apiToken(self):
        return self["auth"]["token"] if ("auth" in self) else None

    @property
    def username(self):
        return self.apiKey

    @property
    def password(self):
        return self.apiToken

    @property
    def credentials(self):
        return (self.apiKey, self.apiToken)

    @property
    def domain(self):
        return self["options"]["domain"]

    @property
    def host(self):
        return self.orgId + "." + self.domain

    @property
    def logLevel(self):
        return self["options"]["logLevel"]

    @property
    def port(self):
        return self["options"]["mqtt"]["port"]

    @property
    def transport(self):
        return self["options"]["mqtt"]["transport"]

    @property
    def cleanStart(self):
        return self["options"]["mqtt"]["cleanStart"]

    @property
    def sessionExpiry(self):
        return self["options"]["mqtt"]["sessionExpiry"]

    @property
    def keepAlive(self):
        return self["options"]["mqtt"]["keepAlive"]

    @property
    def caFile(self):
        return self["options"]["mqtt"]["caFile"]

    @property
    def verify(self):
        return self["options"]["http"]["verify"]


def parseEnvVars():
    """
    Parse environment variables into a Python dictionary suitable for passing to the
    application client constructor as the `options` parameter

    - `WIOTP_IDENTITY_APPID`
    - `WIOTP_AUTH_KEY`
    - `WIOTP_AUTH_TOKEN`
    - `WIOTP_OPTIONS_DOMAIN` (optional)
    - `WIOTP_OPTIONS_LOGLEVEL` (optional)
    - `WIOTP_OPTIONS_MQTT_INSTANCEID` (optional)
    - `WIOTP_OPTIONS_MQTT_PORT` (optional)
    - `WIOTP_OPTIONS_MQTT_TRANSPORT` (optional)
    - `WIOTP_OPTIONS_MQTT_CAFILE` (optional)
    - `WIOTP_OPTIONS_MQTT_CLEANSTART` (optional)
    - `WIOTP_OPTIONS_MQTT_SESSIONEXPIRY` (optional)
    - `WIOTP_OPTIONS_MQTT_KEEPALIVE` (optional)
    - `WIOTP_OPTIONS_HTTP_VERIFY` (optional)
    """

    # Auth
    authKey = os.getenv("WIOTP_AUTH_KEY", None)
    authToken = os.getenv("WIOTP_AUTH_TOKEN", None)

    # Also support WIOTP_API_KEY / WIOTP_API_TOKEN usage
    if authKey is None and authToken is None:
        authKey = os.getenv("WIOTP_API_KEY", None)
        authToken = os.getenv("WIOTP_API_TOKEN", None)

    # Identity
    appId = os.getenv("WIOTP_IDENTITY_APPID", str(uuid.uuid4()))
    # Options
    domain = os.getenv("WIOTP_OPTIONS_DOMAIN", None)
    logLevel = os.getenv("WIOTP_OPTIONS_LOGLEVEL", "info")
    instanceId = os.getenv("WIOTP_OPTIONS_MQTT_INSTANCEID", None)
    port = os.getenv("WIOTP_OPTIONS_MQTT_PORT", None)
    transport = os.getenv("WIOTP_OPTIONS_MQTT_TRANSPORT", None)
    caFile = os.getenv("WIOTP_OPTIONS_MQTT_CAFILE", None)
    cleanStart = os.getenv("WIOTP_OPTIONS_MQTT_CLEANSTART", "True")
    sessionExpiry = os.getenv("WIOTP_OPTIONS_MQTT_SESSIONEXPIRY", "3600")
    keepAlive = os.getenv("WIOTP_OPTIONS_MQTT_KEEPALIVE", "60")
    verifyCert = os.getenv("WIOTP_OPTIONS_HTTP_VERIFY", "True")

    if port is not None:
        try:
            port = int(port)
        except ValueError as e:
            raise ConfigurationException("WIOTP_OPTIONS_MQTT_PORT must be a number")

    try:
        sessionExpiry = int(sessionExpiry)
    except ValueError as e:
        raise ConfigurationException("WIOTP_OPTIONS_MQTT_SESSIONEXPIRY must be a number")

    try:
        keepAlive = int(keepAlive)
    except ValueError as e:
        raise ConfigurationException("WIOTP_OPTIONS_MQTT_KEEPALIVE must be a number")

    if logLevel not in ["error", "warning", "info", "debug"]:
        raise ConfigurationException("WIOTP_OPTIONS_LOGLEVEL must be one of error, warning, info, debug")
    else:
        # Convert log levels from string to int (we need to upper case our strings from the config)
        logLevel = logging.getLevelName(logLevel.upper())

    cfg = {
        "identity": {"appId": appId},
        "options": {
            "domain": domain,
            "logLevel": logLevel,
            "mqtt": {
                "instanceId": instanceId,
                "port": port,
                "transport": transport,
                "cleanStart": cleanStart in ["True", "true", "1"],
                "sessionExpiry": sessionExpiry,
                "keepAlive": keepAlive,
                "caFile": caFile,
            },
            "http": {"verify": verifyCert in ["True", "true", "1"]},
        },
        "auth": {"key": authKey, "token": authToken}
    }

    cfg["auth"] = {"key": authKey, "token": authToken}

    return ApplicationClientConfig(**cfg)


def parseConfigFile(configFilePath):
    """
    Parse a yaml configuration file into a Python dictionary suitable for passing to the
    device client constructor as the `options` parameter

    # Example Configuration File

    identity:
      appId: myApp
    auth:
      key: a-23gh56-sdsdajhjnee
      token: Ab$76s)asj8_s5
    options:
      domain: internetofthings.ibmcloud.com
      logLevel: error|warning|info|debug
      mqtt:
        instanceId: myInstance
        port: 8883
        transport: tcp
        cleanStart: false
        sessionExpiry: 3600
        keepAlive: 60
        caFile: /path/to/certificateAuthorityFile.pem
      http:
        verify: true
    """

    try:
        with open(configFilePath) as f:
            data = yaml.full_load(f)
    except (OSError, IOError) as e:
        # In 3.3, IOError became an alias for OSError, and FileNotFoundError is a subclass of OSError
        reason = "Error reading device configuration file '%s' (%s)" % (configFilePath, e)
        raise ConfigurationException(reason)

    if "options" in data and "logLevel" in data["options"]:
        if data["options"]["logLevel"] not in ["error", "warning", "info", "debug"]:
            raise ConfigurationException("Optional setting options.logLevel must be one of error, warning, info, debug")
        else:
            # Convert log levels from string to int (we need to upper case our strings from the config)
            data["options"]["logLevel"] = logging.getLevelName(data["options"]["logLevel"].upper())

    return ApplicationClientConfig(**data)
